csp_billing_adapter/adapter.py

```diff
@@ -141,60 +143,63 @@ def event_loop_handler(
 ) -> datetime.datetime:
     """Perform the event loop processing actions."""
     log.info('Starting event loop processing')
+    now = get_now()
+    errors = []
 
     try:
         usage = hook.get_usage_data(config=config)
+        log.debug('Retrieved usage data: %s', usage)
     except Exception as exc:
         log.warning("Failed to retrieve usage data: %s", str(exc))
-        now = get_now()
-        hook.update_csp_config(
-            config=config,
-            csp_config={
-                'timestamp': date_to_string(now),
-                'billing_api_access_ok': False,
-                'errors': [
-                    f'Usage data retrieval failed: {exc}'
-                ]
-            },
-            replace=False
-        )
-        # NOTE:
-        # If this happens during initial deployment timeframe, failing
-        # to retrieve usage data here could simple mean that there is
-        # nothing to bill for yet so not proceeding to metering makes
-        # sense.
-        # If this occurs for an existing deployment, with valid usage
-        # records waiting to be billed, continuing on to metering is
-        # worth considering.
-        # However, doing so will currently result in the error state
-        # that was saved in the csp_config being lost, as it would be
-        # overwritten with a success state, or a different error.
+        errors.append(f'Usage data retrieval failed: {exc}')
 
-        log.info('Finishing event loop processing - get_usage_data() error')
-        return now
-
-    log.debug('Retrieved usage data: %s', usage)
-
-    add_usage_record(hook, config, usage)
+    if usage:
+        try:
+            add_usage_record(hook, config, usage)
+        except Exception as error:
+            log.warning("Failed to save usage data to cache: %s", str(error))
+            errors.append(f'Usage data failed to save: {error}')
```

Comment:
So if we get a failure adding a usage record we would still continue on to doing a get_cache(), which could potentially succeed if the add usage error was a write/update phase failure, but the cache was still readable, e.g. read-only access to backing store...

Reply:
Might it make more sense for the add_usage_record() to return the cache, skipping the need to get it here, and in such a case if we failed to successfully add a usage record, we could return an empty dict for cache, and then skip the process_metering() if we had issues updating the cache?

Reply:
Of course in the case where get_usage_data() failed we would skip the add_usage_records(), so we would still need to get_cache()... So maybe we should be doing the get_cache() before the add_usage_record(), passing it into add_usage_record() so that it doesn't need to do so, and then returning the cache from add_usage_record(), which could be an empty cache if things went badly wrong, meaning we would skip the process_metering() step in the event_loop_handler()?

Reply:
Actually it might be useful for the update_cache() hook implementations to return the updated cached, avoiding the need to do a get_cache() if we wanted to get the cache content afterwards...

Reply:
There's also disadvantages to queuing changes and saving all at once. Since all the steps are independent it combines all updates into one call making that one call a single point of failure.

Reply:
Well the code flow as it stands has add_usage_record() calling get_cache() then optionally adding the specified usage record if it isn't a duplicate, followed by the calling routine, event_loop_handler(), doing a get_cache() again... Having add_usage_record() return the (possibly updated) cache on success would eliminate the need to do that second get_cache(), which would be inline with reducing potentially unnecessary get_cache() calls...

As for being a single point of failure, we can potentially incrementally stash all updates to the cache & csp_config, we just need to ensure we don't lose any useful info while doing so, in case the following updates fail... Otherwise we risk loosing useful information, such as the error list, if say a get_usage_data() failed but we had process_metering() raised a hard failure if metering failed...

Reply:
Any optimization I think is a separate task/discussion. First goal is catch and log errors and avoid crash if it's better to keep the adapter running.